File: Problems/0300_0399/0384_Shuffle_an_Array/Project_Go/solution/solution2.go
package solution

import (
	"math/rand"
	"time"
)

// 13ms - 23ms
type Solution2 struct {
	arr []int
}

func Constructor(nums []int) Solution2 {
	rand.Seed(time.Now().UnixNano())
	return Solution2{nums}
}

func (this *Solution2) Reset() []int {
	return this.arr
}

func (this *Solution2) Shuffle() []int {
	arr := make([]int, len(this.arr))
	copy(arr, this.arr)
	for i := 0; i < len(arr)-1; i++ {
		swp_num := rand.Intn(len(arr) - i)
		arr[len(arr)-1-i], arr[swp_num] = arr[swp_num], arr[len(arr)-1-i]
	}
	return arr
}

/*
// Wrong answer 7/8

type Solution struct {
	arr []int
}

func Constructor(nums []int) Solution {
	rand.Seed(time.Now().UnixNano())
	return Solution{nums}
}

func (this *Solution) Reset() []int {
	return this.arr
}

func (this *Solution) Shuffle() []int {
	n := len(this.arr)

	arr := make([]int, len(this.arr))
	copy(arr, this.arr)

	for i := 0; i < n; i++ {
		swp_num := rand.Intn(n)
		arr[0], arr[swp_num] = arr[swp_num], arr[0]
	}
	return arr
}
*/

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.Reset();
 * param_2 := obj.Shuffle();
 */

File: Problems/0300_0399/0384_Shuffle_an_Array/Project_Java/Solution2.java
import java.util.*;

class Solution2 {
    // 48ms - 59ms
    int[] arr;
    public Solution2(int[] nums) {
        arr = nums;
    }
    
    public int[] reset() {
        return arr;
    }
    
    public int[] shuffle() {
        Random rand = new Random();
        int[] ans = arr.clone();
        for (int i = 0; i < ans.length; i++) {
            int swp_num = rand.nextInt(ans.length);
            int temp = ans[i];
            ans[i] = ans[swp_num];
            ans[swp_num] = temp;
        }
        return ans;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */
 